### PR TITLE
Feedback for jhbiggs

### LINKED ISSUE #1581: Remove dagger-android
**Is your feature request related to a problem? Please describe.**
dagger-android adds complication and is no longer being supported as google have decided it is not fit for purpose as announced [last week](https://youtu.be/o-ins1nvbDg?t=539)

**Describe the solution you'd like**
Replace all usages

**Describe alternatives you've considered**
Just continue with the deprecated library

### PR DESCRIPTION
**Pull request by macgills**
Fixes  #1581
@jhbiggs

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -36,6 +37,9 @@ dependencies {
   // use jdk8 java.time backport, as long app < Build.VERSION_CODES.O
   implementation(Libs.threetenabp)
 
+  implementation 'com.google.dagger:dagger:2.25.2'
```
Comment (macgills) [inline core/build.gradle]: you don't need to do this, go to Versions if you want to upgrade dagger

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreComponent.kt [hunk 2]
```diff
@@ -79,6 +79,11 @@ interface CoreComponent {
   fun connectivityManager(): ConnectivityManager
   fun context(): Context
   fun downloader(): Downloader
+  fun bookmarksActivityComponent(): BookmarksActivityComponent.Builder
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreComponent.kt]: We don't actually need all these subcomponents, the dependencies most of these are trying to resolve are all provided by CoreComponent already, you can most likely just add the `inject` methods for the activities and add the modules to the annotation here.

The only concern is if the scoping of the dependencies of the specific modules needs to be different. At that point it makes sense to have a subcomponent, you could also have the dependencies unscoped which would most likely produce the same effect

**Comment by jhbiggs** [comment]
So the app files seem to have fallen into place suspiciously easily after I removed the injection() from BaseActivity.  I am not sure how to ensure functionality of a fragment when injection() is gone as there's no onCreate() function there.  Do I import onCreate()?  Does Dagger somehow know to use the non-overridden injection() function?  Still searching for answers...

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreComponent.kt [hunk 3]
```diff
@@ -52,7 +59,8 @@ import javax.inject.Singleton
     ApplicationModule::class,
     NetworkModule::class,
     JNIModule::class,
-    DataModule::class]
+    DataModule::class,
+    ActivityModule::class]
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreComponent.kt]: The activity module should not be here, activityModule depends on having a component created with an activity dependency provided

**Comment by macgills** [comment]
@jhbiggs you probably don't want to remove injection from baseActivity or more accurately you want to make it abstract and probably add a parameter.

It is very annoying to have to type everywhere `CoreApp.getCoreComponent().inject(this)` when it could be like this and you don't have to worry about calling it in the right place in every class.

```
override fun injection(coreComponent: CoreComponent){
  coreComponent.inject(this)
}
```

We want to construct this in such a way that if we don't inject our dependencies in an activity then the app will refuse to compile. By making `injection()` abstract we say "A subclass of this class must preform injection or it is not fulfilling my contract" and adding the parameter just shunts some common work into the base class.

to make life simpler in app/custom modules I added these extension functions

```
private val BaseActivity.kiwixComponent: KiwixComponent
  get() = (applicationContext as KiwixApp).kiwixComponent

internal inline val BaseActivity.kiwixActivityComponent
  get() = kiwixComponent
    .activityComponentBuilder()
    .activity(this)
    .build()
```

so the call site from kotlin is 'kiwixActivityComponent.inject(this)`


RE: Fragments, there are a couple of sites for injection, `onAttach()` is probably best but I have used `onViewCreated()`.If you want to know what methods are overridable in any class press `ctrl+o`, you will get a full list. Right now we don't scope anything to a fragment particularly but it is important we reuse components so we don't create all of our objects again in a new component. As long as you are only using CoreComponent it is fine because we always retrieve that from the field in the application class but if injection involves an Activity SubComponent then it is a bit more complicated.

**Comment by jhbiggs** [comment]
@macgills could you please expand on your last comment?  Shall I provide a new component for AlertDialogShower for example?

File: core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/BookmarksActivity.java [hunk 4]
```diff
@@ -113,8 +115,9 @@ public void onDestroyActionMode(ActionMode mode) {
 
   @Override
   protected void onCreate(Bundle savedInstanceState) {
-    CoreApp.getCoreComponent().inject(this);
     super.onCreate(savedInstanceState);
+    injection(CoreApp.getCoreComponent());
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/BookmarksActivity.java]: don't call injection, gets called in base class

**Comment by macgills** [comment]
@jhbiggs leave Activity scope on dialog shower, that is provided through an ActivityModule because it needs an activity, the module is bound to an ActivityComponenet that is created with an Activity so it can be provided.
My point was that all these Presenters and the like weren't actually using any activity dependency.

There is also a lot of double/triple injection going on. Each activity should inject once and once only otherwise you are creating objects twice

**Pull request closed without merging**